tests/scripts/test_psa_constant_names.py

```diff
@@ -200,26 +229,35 @@ def parse_header(self, filename):
             for line in lines:
                 self.parse_header_line(line)
 
+    _macro_identifier_re = r'[A-Z]\w+'
+    def generate_undeclared_names(self, expr):
+        for name in re.findall(self._macro_identifier_re, expr):
+            if name not in self.all_declared:
+                yield name
+
+    def accept_test_case_line(self, function, argument):
+        #pylint: disable=unused-argument
+        undeclared = list(self.generate_undeclared_names(argument))
+        if undeclared:
+            raise Exception('Undeclared names in test case', undeclared)
+        return True
+
     def add_test_case_line(self, function, argument):
         """Parse a test case data line, looking for algorithm metadata tests."""
+        sets = []
         if function.endswith('_algorithm'):
-            # As above, ECDH and FFDH algorithms are excluded for now.
-            # Support for them will be added in the future.
-            if 'ECDH' in argument or 'FFDH' in argument:
-                return
-            self.algorithms.add(argument)
-            if function == 'hash_algorithm':
-                self.hash_algorithms.add(argument)
-            elif function in ['mac_algorithm', 'hmac_algorithm']:
-                self.mac_algorithms.add(argument)
-            elif function == 'aead_algorithm':
-                self.aead_algorithms.add(argument)
-        elif function == 'key_type':
-            self.key_types.add(argument)
-        elif function == 'ecc_key_types':
-            self.ecc_curves.add(argument)
-        elif function == 'dh_key_types':
-            self.dh_groups.add(argument)
+            sets.append(self.algorithms)
+            if function == 'key_agreement_algorithm' and \
+               argument.startswith('PSA_ALG_KEY_AGREEMENT('):
+                # We only want *raw* key agreement algorithms as such, so
+                # exclude ones that are already chained with a KDF.
+                # Keep the expression as one to test as an algorithm.
+                function = 'other_algorithm'
+        if function in self.table_by_test_function:
+            sets.append(self.table_by_test_function[function])
```

Comment:
Do we want to call `accept_test_case_line()` if `function` was not in `self.table_by_test_function`? I see `accept_test_case_line()` throws an exception instead of returning `False` for not accepted test cases.

Reply:
Good question. If we add a function to `test_suite_psa_crypto_metadata`, what do we want to happen here? I don't think the current behavior is correct: a new function whose name does not end in `_algorithm` would undergo the argument validity check without contributing to the data tested by this script. I'll change the code to insist on the function being known: better fail with an exception than silently not test something that we might reasonably believe is tested.

Reply:
I've changed the code to insist on knowing the function name. This uncovered a lack of coverage for some expressions involving a wildcard.